engine/runtime/src/main/java/org/enso/interpreter/runtime/ResourceManager.java

```diff
@@ -174,21 +147,82 @@ public void shutdown() {
   }
 
   /**
-   * The worker action for the underlying logic of this module. At least one such thread must be
-   * spawned in order for this module to be operational.
+   * Processes {@link Item}s eligible for GC. Plays two roles. First of all cleans {@link
+   * #referenceQueue} in {@link #run()} method running in its own thread. Then it invokes finalizers
+   * in {@link #perform} method inside of Enso execution context.
    */
-  private class Runner implements Runnable {
+  private final class ProcessItems extends ThreadLocalAction implements Runnable {
+    /**
+     * @GuardedBy("pendingItems")
+     */
+    private final List<Item> pendingItems = new ArrayList<>();
+
+    /**
+     * @GuardedBy("pendingItems")
+     */
+    private Future<Void> request;
+
     private volatile boolean killed = false;
 
+    ProcessItems() {
+      super(false, false, true);
+    }
+
+    /**
+     * Runs at a safe point in middle of regular Enso program execution. Gathers all available
+     * {@link #pendingItems} and runs their finalizers. Removes all processed items from {@link
+     * #pendingItems}. If there are any remaining, continues processing them. Otherwise finishes.
+     *
+     * @param access not used for anything
+     */
+    @Override
+    protected void perform(ThreadLocalAction.Access access) {
+      for (; ; ) {
+        Item[] toProcess;
+        synchronized (pendingItems) {
+          request.cancel(false);
+          if (pendingItems.isEmpty()) {
+            // nothing to process,
+            // signal request is finished
+            request = null;
+            return;
+          }
+          toProcess = pendingItems.toArray(Item[]::new);
+          // mark as being processed
+          pendingItems.set(0, null);
+        }
+        try {
+          for (var it : toProcess) {
+            it.finalizeNow(context);
+            removeFromItems(it);
+          }
+        } finally {
+          synchronized (pendingItems) {
+            pendingItems.subList(0, toProcess.length).clear();
+          }
+        }
+      }
+    }
+
+    /**
+     * Running in its own thread. Waiting for {@link #referenceQueue} to be populated with GCed
+     * items. Scheduling {@link #perform} action at safe points while passing the {@link Item}s to
+     * it via {@link #pendingItems}.
+     */
     @Override
     public void run() {
       while (true) {
         try {
           Reference<? extends ManagedResource> ref = referenceQueue.remove();
           if (!killed) {
             if (ref instanceof Item it) {
-              it.isFlaggedForFinalization().set(true);
-              scheduleFinalizationAtSafepoint(it);
+              it.flaggedForFinalization.set(true);
+              synchronized (pendingItems) {
+                if (request == null) {
+                  request = context.submitThreadLocal(null, this);
```

Comment:
Here is the [submitThreadLocal javadoc](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/TruffleLanguage.Env.html#submitThreadLocal(java.lang.Thread%5B%5D,com.oracle.truffle.api.ThreadLocalAction)).

Reply:
Contiues at the [next PR](https://github.com/enso-org/enso/pull/10665/files#r1691010351).